File: solutions/0665. Non-decreasing Array/0665.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  bool checkPossibility(const vector<int>& nums) {
    int count = 0;
    int prev = INT_MIN;

    for (int i = 0; i < nums.size(); ++i) {
      if (nums[i] < prev) {
        ++count;
        if (count > 1)
          return false;
        if (i > 1 && nums[i - 2] > nums[i])
          continue;
      }
      prev = nums[i];
    }  
    
    return true;
  }
};

File: solutions/0665. Non-decreasing Array/0665.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
  def checkPossibility(self, nums: List[int]) -> bool:
    count = 0
    prev = float('-inf')

    for i in range(len(nums)):
      if nums[i] < prev:
        count += 1
        if count > 1:
          return False
        if i > 1 and nums[i - 2] > nums[i]:
          continue
      prev = nums[i]

    return True    